### PR TITLE
Balance sizes of similar nodegroups in scale-up

### PR DESCRIPTION
**Pull request by MaciekPytel**
Ref: #47

### CODE REVIEW

**Comment by MaciekPytel** [comment]
Please don't merge until all items in the below list are marked as [passed]. Let me know if you have any other ideas for things I should test.

Testing done for this PR:
 - Create cluster with 2 similar node groups and test that scale up is split between those. [passed]
 - Set --balance-similar-node-groups=false, verify that no balancing happens in that case [TODO]
 - Include a node group using different instance type in the same test cluster, check that:
    - Scale up on the 2 similar node groups are not balanced with new node group [passed]
    - Scale up on the new node group is not balanced with the old groups [passed]
 - Removing pods that triggered scale-up triggers scale down normally and everything works as expected [passed]
 - Trigger a scale-up that hits maximum size of one of the groups, verify maximum size is respected [passed]
 - Test that maximum total size of the cluster is respected [TODO]
 - Run e2e test suite with --balance-similar-node-groups=true [IN PROGRESS]
 - Run e2e test suite with --balance-similar-node-groups=false [TODO]

**Comment by MaciekPytel** [comment]
Log sample from balanced scale-up: 
```
I0602 15:45:22.679177       5 scale_up.go:159] Best option to resize: https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2
I0602 15:45:22.679227       5 scale_up.go:163] Estimated 6 nodes needed in https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2
I0602 15:45:22.679299       5 scale_up.go:190] Splitting scale-up between 2 similar node groups: {https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2, https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group}
I0602 15:45:23.076552       5 scale_up.go:201] Final scale-up plan: [{https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2 1->5 (max: 5)} {https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group 3->5 (max: 8)}]
I0602 15:45:23.076614       5 scale_up.go:252] Scale-up: setting group https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2 size to 5
I0602 15:45:23.224068       5 gce_manager.go:134] Setting mig size https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2 to 5
I0602 15:45:23.994176       5 gce_manager.go:147] Waiting for operation maciekpytel-dev-playground us-central1-b operation-1496418323226-550fc0a975d90-17890c4b-9757ad1a
I0602 15:45:24.145755       5 gce_manager.go:149] Operation maciekpytel-dev-playground us-central1-b operation-1496418323226-550fc0a975d90-17890c4b-9757ad1a status: DONE
I0602 15:45:24.145871       5 scale_up.go:252] Scale-up: setting group https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group size to 5
I0602 15:45:24.146350       5 event.go:217] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"978c2a54-47a9-11e7-897d-42010a800002", APIVersion:"v1", ResourceVersion:"1069062", FieldPath:""}): type: 'Normal' reason: 'ScaledUpGroup' Scale-up: group https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group-2 size set to 5
I0602 15:45:24.246657       5 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0602 15:45:24.253822       5 gce_manager.go:134] Setting mig size https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group to 5
I0602 15:45:24.952045       5 gce_manager.go:147] Waiting for operation maciekpytel-dev-playground us-central1-b operation-1496418324255-550fc0aa71119-85a14883-f694e4b3
I0602 15:45:25.056144       5 gce_manager.go:149] Operation maciekpytel-dev-playground us-central1-b operation-1496418324255-550fc0aa71119-85a14883-f694e4b3 status: DONE
I0602 15:45:25.056757       5 event.go:217] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"978c2a54-47a9-11e7-897d-42010a800002", APIVersion:"v1", ResourceVersion:"1069062", FieldPath:""}): type: 'Normal' reason: 'ScaledUpGroup' Scale-up: group https://content.googleapis.com/compute/v1/projects/maciekpytel-dev-playground/zones/us-central1-b/instanceGroups/e2e-test-maciekpytel-minion-group size set to 5
```

**Comment by MaciekPytel** [comment]
cc: @mumoshu

File: cluster-autoscaler/core/scale_up.go [hunk 1]
```diff
@@ -154,48 +162,53 @@ func ScaleUp(context *AutoscalingContext, unschedulablePods []*apiv1.Pod, nodes
...
-			glog.V(1).Infof("Capping size to MAX (%d)", bestOption.NodeGroup.MaxSize())
-			newSize = bestOption.NodeGroup.MaxSize()
-		}
-
-		if context.MaxNodesTotal > 0 && len(nodes)+(newSize-currentSize) > context.MaxNodesTotal {
+		if context.MaxNodesTotal > 0 && len(nodes)+bestOption.NodeCount > context.MaxNodesTotal {
 			glog.V(1).Infof("Capping size to max cluster total size (%d)", context.MaxNodesTotal)
-			newSize = context.MaxNodesTotal - len(nodes) + currentSize
-			if newSize < currentSize {
+			bestOption.NodeCount = context.MaxNodesTotal - len(nodes)
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_up.go]: do we have to modify the bestOption

File: cluster-autoscaler/core/scale_up.go [hunk 2]
```diff
@@ -154,48 +162,53 @@ func ScaleUp(context *AutoscalingContext, unschedulablePods []*apiv1.Pod, nodes
...
+				for _, ng := range targetNodeGroups {
+					buffer.WriteString(", ")
+					buffer.WriteString(ng.Id())
+				}
+				glog.V(1).Infof("Splitting scale-up between %v similar node groups: {%v}", len(targetNodeGroups)+1, buffer.String())
+			}
+		} else {
+			targetNodeGroups = make([]cloudprovider.NodeGroup, 0)
+		}
+		targetNodeGroups = append(targetNodeGroups, bestOption.NodeGroup)
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_up.go]: move it before if

File: cluster-autoscaler/core/scale_up.go (hunk 1 above)
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_up.go]: done (created a local variable instead)

File: cluster-autoscaler/core/scale_up.go (hunk 2 above)
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_up.go]: done

**Pull request merged**